### PR TITLE
MueLu coarse coordinates

### PR DESCRIPTION
**Pull request by lucbv**
@trilinos/muelu 

## Description
Modifying the creation of coarse coordinates from CoordinateTransferFactory to prolongators.

## Motivation and Context
This avoids having multiple intricate algorithm in the transfer factory when the prolongators already implement most of the work.

## Related Issues

* Blocks #2577 
* Follows #2362 
* Part of #2230  #2687 

## How Has This Been Tested?
All unit-test requiring coordinates to be injected in MueLu are still passing

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.

### CODE REVIEW

**Comment by cgcgcg** [comment]
@mhoemmen Yes, you could. The idea was to have one single typedef for a real-valued multivector, instead of having it in every single file. In the unlikely scenario that we want to change the type for coordinates, we have to edit a single file, instead of grepping through everything for hard-coded doubles..

File: packages/muelu/src/Interface/MueLu_MLParameterListInterpreter_def.hpp [hunk 1]
```diff
@@ -256,14 +285,30 @@ namespace MueLu {
...
+        CoupledAggFact2->SetMaxNeighAlreadySelected(maxNbrAlreadySelected);
+        CoupledAggFact2->SetOrdering("natural");
+        CoupledAggFact2->SetPhase3AggCreation(0.5);
+        CoupledAggFact2->SetFactory("Graph", dropFact);
+        CoupledAggFact2->SetFactory("DofsPerNode", dropFact);
+        AggFact = CoupledAggFact2;
+      }
+#else
+      RCP<CoupledAggregationFactory> CoupledAggFact2 = rcp( new CoupledAggregationFactory() );
+      CoupledAggFact2 = rcp( new CoupledAggregationFactory() );
```
Comment (cgcgcg) [inline packages/muelu/src/Interface/MueLu_MLParameterListInterpreter_def.hpp]: I guess this is superfluous?

**Comment by mhoemmen** [comment]
@cgcgcg So why not have a typedef just for the type of coordinates, instead of having a typedef for the type of a MultiVector containing coordinates?  A typedef just for the type of a coordinate would be  orthogonal to the other template parameters.

File: packages/muelu/src/Interface/MueLu_MLParameterListInterpreter_def.hpp [hunk 2]
```diff
@@ -84,6 +84,16 @@
 #include "MueLu_NullspaceFactory.hpp"
 #include "MueLu_ParameterListUtils.hpp"
 
+#ifdef HAVE_MUELU_KOKKOS_REFACTOR
+#include "MueLu_CoalesceDropFactory_kokkos.hpp"
+// #include "MueLu_CoarseMapFactory_kokkos.hpp"
+// #include "MueLu_CoordinatesTransferFactory_kokkos.hpp"
+// #include "MueLu_NullspaceFactory_kokkos.hpp"
```
Comment (cgcgcg) [inline packages/muelu/src/Interface/MueLu_MLParameterListInterpreter_def.hpp]: How come we don't need NullspaceFactory_kokkos, but the regular NullspaceFactory?

**Comment by cgcgcg** [comment]
@mhoemmen True, we could have done that. I prefer this, since we like to use the shorthand `MultiVector`, and this then just means we can use `RealValuedMultiVector` without any template parameters.

**Comment by mhoemmen** [comment]
Would you ever want to support multiple coordinate types in a single MueLu build?  If not, why not use a typedef instead of a template parameter?

Compare: Tpetra has a LocalOrdinal type template parameter, but it's only ever been tested with `int` (with very rare exceptions).  It should be (and will become) just a typedef `local_ordinal_type`.

**Comment by lucbv** [comment]
@mhoemmen, I hear you but honestly based on how this was handled so far I do not feel confident that people will remember to do typedef appropriately, the RealValuedMultiVector seems easier to uniformize the handling of coordinates even if it is not the most elegant.
I will also look into modifying the shorthand to have it templated on `Teuchos::ScalarTraits<Scalar>::magnitudeType` so that we never encounter undefined behavior when coordinates could of type `double` and scalar of type `float` or `quad`?

**Comment by mhoemmen** [comment]
@lucbv I would like to push back a little bit on this.

Intel debug builds literally don't work for SPARC any more because the linker overflows `unsigned int`.  To be blunt, if we (Sierra and ATDM apps) can't build MueLu, we won't use it.

The lesson learned in Tpetra is that if you give users a template parameter, they will put anything in there that compiles.  GlobalOrdinal should never have been a template parameter.  Now we find ourselves with `Tpetra_INST_INT_*` for `* = INT, LONG, LONG_LONG, UNSIGNED_LONG`.  This adds almost no value to users, while forcing Trilinos to build Tpetra and all downstream packages over again for each supported GlobalOrdinal type.

If you expose a new template parameter without constraining it, you will find yourself supporting many more possible template parameter values than you intended.

Literally this is how I would fix this problem:
```c++

// Type of a coordinate
using coordinate_type = double;

// Type of a MultiVector of coordinates
`template<class MultiVectorType>
using coordinate_multivector_type = Xpetra::MultiVector<coordinate_type, typename MultiVectorType::local_ordinal_type, typename MultiVectorType::global_ordinal_type, typename MultiVectorType::node_type>;
```
Now both users and MueLu developers can do this:
```c++
using XMV = Xpetra::MultiVector<Scalar, LO, GO, Node>;
XMV X (...);
coordinate_multivector_type<XMV> coords (...);
```

**Comment by lucbv** [comment]
@mhoemmen, I am working on this, the level of pain is at about 120%, I hope this will decrease our build time/space by a reasonable amount. 
Also ironically even though I am removing this template parameter `Galeri::Xpetra::BuildProblem` is still templated on `SC,LO,GO,Map,CrsMatrixWrap,MultiVector` which seems redundant since `MultiVector` has `typedefs` for `SC,LO,GO`...

**Comment by mhoemmen** [comment]
@lucbv wrote:
> the level of pain is at about 120%

That's not so bad for this solver stack ;-)

> I hope this will decrease our build time/space by a reasonable amount.

Me too, though I think it's still worth having orthogonal template parameters, even if there is no build time or size improvement.

> Galeri::Xpetra::BuildProblem is still templated on SC,LO,GO,Map,CrsMatrixWrap,MultiVector which seems redundant since MultiVector has typedefs for SC,LO,GO...

I agree that we should change this.  I filed an issue for this:  https://github.com/trilinos/Trilinos/issues/3077

**Comment by lucbv** [comment]
@mhoemmen can you look at this again, I think this satisfies your requirement now, It also passes tests, so it should be ready to be merge.

File: packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp [hunk 3]
```diff
@@ -61,6 +61,7 @@ namespace Galeri {
     template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Map, typename Matrix, typename MultiVector>
     class Elasticity2DProblem : public Problem<Map,Matrix,MultiVector> {
     public:
+      using RealValuedMultiVector = typename Problem<Map,Matrix,MultiVector>::RealValuedMultiVector;
```
Comment (mhoemmen) [inline packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp]: I much prefer this -- thanks!

File: packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp [hunk 4]
```diff
@@ -313,15 +314,18 @@ namespace Galeri {
...
       // as we cannot construct a single DOF map in Problem, we repeat the coords
-      this->Coords_ = MultiVectorTraits<Map,MultiVector>::Build(this->Map_, nDim_);
+      this->Coords_ = MultiVectorTraits<Map,RealValuedMultiVector>::Build(this->Map_, nDim_);
 
+      typedef typename RealValuedMultiVector::scalar_type real_type;
       typedef Teuchos::ScalarTraits<Scalar> TST;
 
-      Teuchos::ArrayRCP<SC> x = this->Coords_->getDataNonConst(0);
-      Teuchos::ArrayRCP<SC> y = this->Coords_->getDataNonConst(1);
+      Teuchos::ArrayRCP<real_type> x = this->Coords_->getDataNonConst(0);
```
Comment (mhoemmen) [inline packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp]: For later (not for this PR), replace use of `getData` and `getDataNonConst` to use Kokkos data structures.

File: packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp [hunk 5]
```diff
@@ -370,8 +375,8 @@ namespace Galeri {
       }
 
       // Calculate center
-      SC cx = this->Coords_->getVector(0)->meanValue();
-      SC cy = this->Coords_->getVector(1)->meanValue();
+      real_type cx = this->Coords_->getVector(0)->meanValue();
```
Comment (mhoemmen) [inline packages/galeri/src-xpetra/Galeri_Elasticity2DProblem.hpp]: For later (not this PR), consider merging these two all-reduces into one.

File: packages/galeri/src-xpetra/Galeri_HelmholtzFEM2DProblem.hpp [hunk 6]
```diff
@@ -61,12 +61,12 @@ namespace Galeri {
 
   namespace Xpetra {
 
-    template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Map, typename Matrix, typename MultiVector>
-    class HelmholtzFEM2DProblem : public Problem_Helmholtz<Map,Matrix,MultiVector> {
+    template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Map, typename Matrix, typename MultiVector, typename RealValuedMultiVector>
```
Comment (mhoemmen) [inline packages/galeri/src-xpetra/Galeri_HelmholtzFEM2DProblem.hpp]: Why does HelmholtzFEM2DProblem still take a RealValuedMultiVector template parameter?

File: packages/galeri/src-xpetra/Galeri_StencilProblems_Helmholtz.hpp [hunk 7]
```diff
@@ -57,16 +57,16 @@ namespace Galeri {
   namespace Xpetra {
 
     // =============================================  Helmholtz1D  =============================================
-    template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Map, typename Matrix, typename MultiVector>
-    class Helmholtz1DProblem : public Problem_Helmholtz<Map,Matrix,MultiVector> {
+    template <typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Map, typename Matrix, typename MultiVector, typename RealValuedMultiVector>
```
Comment (mhoemmen) [inline packages/galeri/src-xpetra/Galeri_StencilProblems_Helmholtz.hpp]: Why does HelmholtzFEM1DProblem still take a RealValuedMultiVector template parameter?

File: packages/galeri/src-xpetra/Galeri_HelmholtzFEM2DProblem.hpp (hunk 6 above)
Comment (lucbv) [reply packages/galeri/src-xpetra/Galeri_HelmholtzFEM2DProblem.hpp]: Good catch, this only gets compiled with `MueLu_ENABLE_Experimental=ON`, I'll take care of this.

File: packages/galeri/src-xpetra/Galeri_StencilProblems_Helmholtz.hpp (hunk 7 above)
Comment (lucbv) [reply packages/galeri/src-xpetra/Galeri_StencilProblems_Helmholtz.hpp]: Same as above

**Comment by lucbv** [comment]
@mhoemmen, alright I modified the Helmholtz tests and I did a build with `MueLu_ENABLE_Experimental=ON` and a build with `COMPLEX=ON`. No problem with either of them so this should be good to go now!

**Pull request merged**